oil_lang/expr_eval.py

```diff
@@ -866,70 +1032,145 @@ def _EvalSlice(self, node):
+        if left.tag() != right.tag():
+          result = False
+        else:
+          result = _ValuesEqual(left, right)
       elif op.id == Id.Expr_NotDEqual:
-        result = left != right
+        if left.tag() != right.tag():
+          result = True
+        else:
+          result = not _ValuesEqual(left, right)
 
       elif op.id == Id.Expr_In:
-        result = left in right
+        result = _ValueContains(left, right)
       elif op.id == Id.Node_NotIn:
-        result = left not in right
+        result = not _ValueContains(left, right)
 
       elif op.id == Id.Expr_Is:
+        if left.tag() != right.tag():
+          raise error.InvalidType('Mismatched types', op)
+
         result = left is right
       elif op.id == Id.Node_IsNot:
+        if left.tag() != right.tag():
+          raise error.InvalidType('Mismatched types', op)
+
         result = left is not right
 
       elif op.id == Id.Expr_DTilde:
         # no extglob in Oil language; use eggex
-        return libc.fnmatch(right, left)
+        if left.tag() != value_e.Str:
+          raise error.InvalidType('LHS must be Str', op)
+
+        if right.tag() != value_e.Str:
+          raise error.InvalidType('RHS must be Str', op)
+
+        UP_left = left
+        UP_right = right
+        left = cast(value.Str, UP_left)
+        right = cast(value.Str, UP_right)
+        return value.Bool(libc.fnmatch(right.s, left.s))
+
       elif op.id == Id.Expr_NotDTilde:
-        return not libc.fnmatch(right, left)
+        if left.tag() != value_e.Str:
+          raise error.InvalidType('LHS must be Str', op)
+
+        if right.tag() != value_e.Str:
+          raise error.InvalidType('RHS must be Str', op)
+
+        UP_left = left
+        UP_right = right
+        left = cast(value.Str, UP_left)
+        right = cast(value.Str, UP_right)
+        return value.Bool(not libc.fnmatch(right.s, left.s))
 
       elif op.id == Id.Expr_TildeDEqual:
         # Approximate equality
-        if not isinstance(left, str):
+        UP_left = left
+        if left.tag() != value_e.Str:
           e_die('~== expects a string on the left', op)
 
-        left = left.strip()
-        if isinstance(right, str):
-          return left == right
+        left = cast(value.Str, UP_left)
+        left2 = left.s.strip()
+
+        UP_right = right
+        with tagswitch(right) as case:
+          if right.tag() == value_e.Str:
+            right = cast(value.Str, UP_right)
+            return value.Bool(left2 == right.s)
 
-        if isinstance(right, bool):  # Python quirk: must come BEFORE int
-          left = left.lower()
-          if left in ('true', '1'):
-            left2 = True
-          elif left in ('false', '0'):
-            left2 = False
-          else:
-            return False
+          elif case(value_e.Bool):
+            right = cast(value.Bool, UP_right)
+            left2 = left2.lower()
+            lb = False
+            if left2 in ('true', '1'):
```

Comment:
random note that doesn't relate to this PR -- I wonder if we should not equate `true` and `1`, like in C

Because it's actually the opposite in shell!   e.g. `if grep foo bar` takes the branch if grep returns 0!